src/fee_estimator.rs

```diff
@@ -198,3 +148,66 @@ where
 		self.estimate_fee_rate(confirmation_target.into()).to_sat_per_kwu() as u32
 	}
 }
+
+pub(crate) fn get_num_block_defaults_for_target(target: ConfirmationTarget) -> usize {
+	match target {
+		ConfirmationTarget::OnchainPayment => 6,
+		ConfirmationTarget::ChannelFunding => 12,
+		ConfirmationTarget::Lightning(ldk_target) => match ldk_target {
+			LdkConfirmationTarget::MaximumFeeEstimate => 1,
+			LdkConfirmationTarget::UrgentOnChainSweep => 6,
+			LdkConfirmationTarget::MinAllowedAnchorChannelRemoteFee => 1008,
+			LdkConfirmationTarget::MinAllowedNonAnchorChannelRemoteFee => 144,
+			LdkConfirmationTarget::AnchorChannelFee => 1008,
+			LdkConfirmationTarget::NonAnchorChannelFee => 12,
+			LdkConfirmationTarget::ChannelCloseMinimum => 144,
+			LdkConfirmationTarget::OutputSpendingFee => 12,
+		},
+	}
+}
+
+pub(crate) fn get_fallback_rate_for_target(target: ConfirmationTarget) -> u32 {
+	match target {
+		ConfirmationTarget::OnchainPayment => 5000,
+		ConfirmationTarget::ChannelFunding => 1000,
+		ConfirmationTarget::Lightning(ldk_target) => match ldk_target {
+			LdkConfirmationTarget::MaximumFeeEstimate => 8000,
+			LdkConfirmationTarget::UrgentOnChainSweep => 5000,
+			LdkConfirmationTarget::MinAllowedAnchorChannelRemoteFee => FEERATE_FLOOR_SATS_PER_KW,
+			LdkConfirmationTarget::MinAllowedNonAnchorChannelRemoteFee => FEERATE_FLOOR_SATS_PER_KW,
+			LdkConfirmationTarget::AnchorChannelFee => 500,
+			LdkConfirmationTarget::NonAnchorChannelFee => 1000,
+			LdkConfirmationTarget::ChannelCloseMinimum => 500,
+			LdkConfirmationTarget::OutputSpendingFee => 1000,
+		},
+	}
+}
+
+pub(crate) fn get_all_conf_targets() -> [ConfirmationTarget; 10] {
+	[
+		ConfirmationTarget::OnchainPayment,
+		ConfirmationTarget::ChannelFunding,
+		LdkConfirmationTarget::MaximumFeeEstimate.into(),
+		LdkConfirmationTarget::UrgentOnChainSweep.into(),
+		LdkConfirmationTarget::MinAllowedAnchorChannelRemoteFee.into(),
+		LdkConfirmationTarget::MinAllowedNonAnchorChannelRemoteFee.into(),
+		LdkConfirmationTarget::AnchorChannelFee.into(),
+		LdkConfirmationTarget::NonAnchorChannelFee.into(),
+		LdkConfirmationTarget::ChannelCloseMinimum.into(),
+		LdkConfirmationTarget::OutputSpendingFee.into(),
+	]
+}
+
+pub(crate) fn apply_post_estimation_adjustments(
+	target: ConfirmationTarget, estimated_rate: FeeRate,
+) -> FeeRate {
+	match target {
+		ConfirmationTarget::Lightning(
+			LdkConfirmationTarget::MinAllowedNonAnchorChannelRemoteFee,
+		) => {
+			let slightly_less_than_background = estimated_rate.to_sat_per_kwu() - 250;
+			FeeRate::from_sat_per_kwu(slightly_less_than_background)
+		},
+		_ => estimated_rate,
+	}
```

Comment:
Will these helpers be called more than once? Just wondering what the rationale for splitting the logic across files. If the only difference later is how `convert_fee_rate` is called, seems that `update_fee_estimates` could be parameterized by a trait instead.

Otherwise, all the fee estimation logic is in the `chain` module instead of here, which seems wrong. A similar argument can be made for other code moved to `chain`, IMO (e.g., broadcasting).

Reply:
Hmm... right, but it seems doable. For instance, the trait's method could take a list of confirmation targets and return an iterator over fee rates. This would keep the code fairly DRY, avoiding repeating much of the logic and keeping these functions internal to the `fee_estimation` module, IIUC.

Reply:
Yeah, an internal enum but I believe it could be implemented on `EsploraAsyncClient`. You wouldn't need individual `ChainSource` objects.

Reply:
Hmm... why couldn't the trait be internal to LDK Node?

Reply:
To be clear, all I'm suggesting is parameterizing `OnchainFeeEstimator::update_fee_estimates` by a trait that can be implemented for both `EsploraAsyncClient` and `BitcoindRpcClient`. Then we wouldn't need to move all the code just to move it back later.

Reply:
Hmm... or possibly returning some sort of async iterator? Alternatively, `OnchainFeeEstimator::update_fee_estimates` can be made into a macro and instantiated for each.

Reply:
Right, I was thinking something like this:

```rust
pub(crate) async fn update_fee_estimates<F: FeeSource>(&self, fee_source: f) -> Result<(), Error> {
    let mut new_fee_rate_cache = HashMap::with_capacity(10);
    let confirmation_targets = vec![/* .. */];
    for (target, fee_rate) in fee_source.fee_rates(confirmation_targets) {
        let adjusted_fee_rate = /* .. */;
        new_fee_rate_cache.insert(target, adjusted_fee_rate);
    }

    if fee_estimator.set_fee_rate_cache(new_fee_rate_cache) { /* .. */ }
}
```

Where `FeeSource::fee_rates` returns an iterator. I'm handwaving the async part -- so fair enough if our MSRV prevents this -- but I'd think the different logic could be encapsulated in `FeeSource::fee_rates` implementations since it is given all targets. For esplora, it can fetch all at once using `get_fee_estimates`  -- so the trait method would need to be `async` as you said -- and return an iterator callinh `convert_fee_rate` on each item. For bitcoind, it can return an iterator that fetches them one-by-one.

Maybe there is a way to do it with a `Fn` trait parameter instead if using `aync_trait` is a problem? Something like: https://forum.dfinity.org/t/passing-async-function-or-closure-as-parameter/21785/9

Anyhow, happy to leave at is for now given the difficulties with `async` / blocking.